From b0e8a008dbab80f550f06b297b017c59dff5abd8 Mon Sep 17 00:00:00 2001
From: Axel Rutz <axel.rutz@clever-systems.net>
Date: Thu, 16 Jul 2015 02:51:02 +0200
Subject: [PATCH] Issue #1377286: Translate label of DS field

---
 ds.ds_fields_info.inc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/ds.ds_fields_info.inc b/ds.ds_fields_info.inc
index 34958e2..a36a22d 100644
--- a/ds.ds_fields_info.inc
+++ b/ds.ds_fields_info.inc
@@ -19,7 +19,8 @@ function ds_ds_fields_info($entity_type) {
   foreach ($custom_fields as $key => $field) {
     if (isset($field->entities[$entity_type])) {
       $fields[$entity_type][$key] = array(
-        'title' => $field->label,
+        // @fixme Use i18n instead.
+        'title' => t($field->label),
         'field_type' => $field->field_type,
         'properties' => $field->properties,
       );
-- 
2.1.0

